drivers: flash: bug fix stm32 ospi flash erase

tu-brian opened the pull request:
This commit fixes a bug in the STM32 ospi flash driver when attempting to erase an area that spans more than one erase sector. Without this fix, only the first sector is actually erased, the rest silently fails the erase.
Issue is that the write enable latch command is only sent for the first erase command.

samples/drivers/spi_flash is updated to check for this condition.

Tested with: west build -b b_u585i_iot02a samples/drivers/spi_flash

Signed-off-by: Brian Juel Folkmann <bju@trackunit.com>

erwango commented:
Hit the wrong button. Not reviewed yet.

FRASTM commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -952,6 +946,13 @@ static int flash_stm32_ospi_erase(const struct device *dev, off_t addr,
+		if (stm32_ospi_write_enable(&dev_data->hospi,
+			dev_cfg->data_mode, dev_cfg->data_rate) != 0) {
+			LOG_ERR("Erase failed : write enable");
+			return -EIO;
```
Could you check what is the ospi_lock_thread/ospi_unlock_thread  position when  we are returning with error in the middle of the while loop. 
Before this change, I  agree that ospi_lock_thread(dev); should have been  placed _after_ the stm32_ospi_mem_ready
but inside the while loop, using break and ret code will avoid returning without  ospi_unlock_thread()

Note that during flash_stm32_ospi_write, ospi_lock_thread should also be after  stm32_ospi_mem_ready()

FRASTM replied:
> Feels like you are doing one-too many things as part of a single commit, 

Especially changing the sample is a good idea of enhancement, because it helps demonstrating the bug/correction.
I agree that this is at least another commit, as it affects other than stm32 driver. 
This change (more sectors) could be proposed as a new items or options in the sample.

tu-brian replied:
So, I should revert the changes in the sample and introduce a new function in the sample that test multi sector erase protect this as compile option. Then split the pr into the bug itself and the change in the sample. 
Is this correctly understood?

DhruvaG2000 reviewed (commented):
In my humble opinion the commit message need not have,
> Tested with: west build -b b_u585i_iot02a samples/drivers/spi_flash

In general commit message dont have tested with ... Commit should explain just what the change is doing and that should be enough.

> samples/drivers/spi_flash is updated to check for this condition.

We can see with git stat which file is being edited, I think this too can be dropped.

Otherwise the commit message is fine. I am yet to review the code changes.

DhruvaG2000 reviewed (changes requested):
Feels like you are doing one-too many things as part of a single commit, please see if you can split up these changes.
In my humble opinion we can split up the changes in sample and driver into 2 commits.
I would even split the sample changes because you are doing multiple things in the sample itself, reading, erasing and writing the flash.
I would strongly suggest splitting up these changes into 2-3 commit messages suiting each change.
Also note that each commit should independently build and be functional, so split it keeping this in mind. I dont see much new additions like change of API's so this should not be too difficult a task.

FRASTM commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -952,6 +947,14 @@ static int flash_stm32_ospi_erase(const struct device *dev, off_t addr,
 
 	while ((size > 0) && (ret == 0)) {
+
+		if (stm32_ospi_write_enable(&dev_data->hospi,
```
Does the code of the flash_stm32_ospi_write() function fits here :

```suggestion
		ret = stm32_ospi_write_enable(&dev_data->hospi,
						    dev_cfg->data_mode, dev_cfg->data_rate);
		if (ret != 0) {
			LOG_ERR("Erase failed : write enable");
			break;
		```

tu-brian replied:
Thanks, I will update this

FRASTM commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -1189,6 +1199,7 @@ static int flash_stm32_ospi_write(const struct device *dev, off_t addr,
 		ret = stm32_ospi_write_enable(&dev_data->hospi,
 						    dev_cfg->data_mode, dev_cfg->data_rate);
 		if (ret != 0) {
+			ospi_unlock_thread(dev);
```
Is it really needed ?
 when breaking the  
`ospi_unlock_thread(dev);`  comes just before the 	`return ret;`   out of the while loop

tu-brian replied:
I will fix this

tu-brian committed:
drivers: flash: bug fix stm32 ospi flash erase

This commit fixes a bug in the STM32 ospi flash driver when attempting
to erase an area that spans more than one erase sector.
Without this fix, only the first sector is actually erased, the rest
silently fails the erase.
Issue is that the write enable latch command is only sent for the first
erase command.

Signed-off-by: Brian Juel Folkmann <bju@trackunit.com>

tu-brian committed:
samples: spi_flash: Add option to test multiple sectors

Add tests for erase of multiple consequtive sectors.

Signed-off-by: Brian Juel Folkmann <bju@trackunit.com>

erwango reviewed (changes requested):
> samples/drivers/spi_flash is updated to check for this condition.

Samples are mainly meant to demonstrate use of subsystems (flash drivers in this case).
If you want to test specific conditions, modify `tests/drivers/flash`.

tu-brian commented:
> > samples/drivers/spi_flash is updated to check for this condition.
> 
> Samples are mainly meant to demonstrate use of subsystems (flash drivers in this case). If you want to test specific conditions, modify `tests/drivers/flash`.

The tests in 'tests/drivers/flash' tests the internal flash of the MCU. I really don't want to change this - but at the same time have an option to test my modifications

erwango commented:
> The tests in 'tests/drivers/flash' tests the internal flash of the MCU. I really don't want to change this - but at the same time have an option to test my modifications

Then we might need dedicated "SPI Flash" tests. But I stand on the position that samples should not be used for extensive testing.

tu-brian commented:
I respect your position. 

That being said, creating a new testsuite for an otherwise totally untested driver is imo way out of scope for submitting a bugfix. 

My objective with this PR was is to fix a bug that we spent quite some time to narrow down, so that others would not have to spent the same effort.  I tried to be a nice guy and updated the sample to prove that the change I did worked. 

I have already spent way much more time on submitting this issue than I can afford, so I now see 3 options:
1) You accept the PR as is (yes, I will fix the signed off by issue)
2) You accept the PR without changes to the sample and then there are no way to verify this in the future
3) 2I withdraw the PR and maintain the fix in our own fork. Then the next person will need to debug and fix this once again.

Option 3 would nor really improve the quality of of zephyr going forward. 

Please let me know which option you chose.

erwango commented:
@tu-brian The real issue in current case is that there is no real spi flash test suite.
We both agree this is unfortunate that there will be no way to verify this in future but we need to draw some boundaries to the modifications that are done to samples.
This being said, I would be ok to accept sample update if I know this will be converted to a test suite eventually. An issue should be raised to follow this up. @de-nordic What do you think?

de-nordic commented:
> @tu-brian The real issue in current case is that there is no real spi flash test suite. We both agree this is unfortunate that there will be no way to verify this in future but we need to draw some boundaries to the modifications that are done to samples. This being said, I would be ok to accept sample update if I know this will be converted to a test suite eventually. An issue should be raised to follow this up. @de-nordic What do you think?

I am willing to accept the sample test, although I am not fine with making multi-sector test dependent on the CONFIG_FLASH_STM32_OSPI, because the sample should have generic behaviour for all SPI boards. 

@erwango Regarding lack of SPI flash tests I will log an Issue with Zephyr for the flash subsystem, as an enhancement, and will make it 3.4 target to have SPI flash test available. This sample is already unfortunate to use "test" in some of the messages, so it's could be misunderstood to serve dual purpose.

tu-brian commented:
What Can I do to move this issue forward?

de-nordic commented:
> What Can I do to move this issue forward?

One question, why the flash tests, that are here https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/flash, can not be altered to do the testing here?
I do not think that we need to have SPI flash specific flash tests, because they use the same Flash API as any other flash access.
Making the test work with Flash API is just the matter of pointing to a device hanging from SPI and adding the test cases for the boundary erase.

tu-brian commented:
I tried to enable the tests https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/flash for the board in question - only to find that the only test present failed, due to unaligned access not being support for the flash device in question. 
At that time I could not afford to spend any more time on this issue, so my previous comments still stands:

I have already spent way much more time on submitting this issue than I can afford, so I now see 3 options:

1.  You accept the PR as is (yes, I will fix the signed off by issue)
2.  You accept the PR without changes to the sample and then there are no way to verify this in the future
3. I withdraw the PR and maintain the fix in our own fork. Then the next person will need to debug and fix this once again.

Option 3 would nor really improve the quality of of zephyr going forward.

Please let me know which option you chose.

de-nordic reviewed (dismissed):
Looks OK.
Approving with sample used to test the fix. I will log an issue for flash drivers/tests to extend tests to cover more scenarios including the one tested by sample.

fabiobaltieri commented:
Hey there's a couple of missing CI tests, I'm kicking a rebase to retrigger them.
